firewood/src/proof.rs

```diff
@@ -455,93 +447,50 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
-                let cur_key = cur_key_path.into_inner();
-
-                let data: Vec<u8> = items.next().unwrap().decode()?;
-
+        let node = NodeType::decode(buf)?;
+        let new_node = merkle
+            .new_node(Node::new(node))
+            .map_err(ProofError::InvalidNode)?;
+        let addr = new_node.as_ptr();
+        match new_node.inner() {
+            NodeType::Leaf(n) => {
+                let cur_key = &n.path().0;
                 // Check if the key of current node match with the given key.
-                if key.len() < cur_key.len() || key[..cur_key.len()] != cur_key {
-                    let ext_ptr = get_ext_ptr(merkle, term, Data(data), CurKey(cur_key))?;
-
-                    return Ok((ext_ptr, None, 0));
+                if key.len() < cur_key.len() || &key[..cur_key.len()] != cur_key {
+                    return Ok((addr, None, 0));
                 }
 
-                let subproof = if term {
-                    Some(SubProof {
-                        encoded: data.clone(),
-                        hash: None,
-                    })
-                } else {
-                    generate_subproof(data.clone()).map(Some)?
-                };
-
-                let cur_key_len = cur_key.len();
-
-                let ext_ptr = get_ext_ptr(merkle, term, Data(data), CurKey(cur_key))?;
-
-                Ok((ext_ptr, subproof, cur_key_len))
+                let subproof = Some(SubProof {
+                    encoded: n.data().to_vec(),
+                    hash: None,
+                });
+                Ok((addr, subproof, cur_key.len()))
             }
+            NodeType::Extension(n) => {
+                let cur_key = &n.path().0;
 
-            BRANCH_NODE_SIZE => {
-                // we've already validated the size, that's why we can safely unwrap
-                let data = items.pop().unwrap().decode()?;
-                // Extract the value of the branch node and set to None if it's an empty Vec
-                let value = Some(data).filter(|data| !data.is_empty());
-
-                // Record encoded values of all children.
-                let mut chd_encoded: [Option<Vec<u8>>; NBRANCH] = Default::default();
-
-                // we popped the last element, so their should only be NBRANCH items left
-                for (i, chd) in items.into_iter().enumerate() {
-                    let data = chd.decode()?;
-                    chd_encoded[i] = Some(data).filter(|data| !data.is_empty());
-                }
-
-                // If the node is the last one to be decoded, then no subproof to be extracted.
-                if end_node {
-                    let branch_ptr = build_branch_ptr(merkle, value, chd_encoded)?;
-
-                    return Ok((branch_ptr, None, 1));
+                // Check if the key of current node match with the given key.
+                if key.len() < cur_key.len() || &key[..cur_key.len()] != cur_key {
```

Comment:
Nit, repeated logic should live in a function. If the function is well-named, we wouldn't need a code comment. 

```rust
pub trait ContainsOtherExt {
    fn contains_other(&self, other: &Self) -> bool;
}

impl<T> ContainsOtherExt for &[T] {
    fn contains_other(&self, other: &Self) -> bool {
        self.len() >= other.len() && &self[..other.len()] == other
    }
}
```
Then the code reads like English
```rust
if !key.contains_other(cur_key) {
// ...
```